barretenberg/cpp/src/barretenberg/vm/avm/trace/gadgets/range_check.hpp

```diff
@@ -0,0 +1,133 @@
+#pragma once
+
+#include "barretenberg/vm/avm/generated/relations/range_check.hpp"
+#include "barretenberg/vm/avm/trace/common.hpp"
+#include <cstdint>
+
+enum class EventEmitter {
+    ALU,
+    MEMORY,
+};
+
+namespace bb::avm_trace {
+// template <class TraceUtilisingRange>
+class AvmRangeCheckBuilder {
+  public:
+    struct RangeCheckEvent {
+        uint32_t clk;
+        FF value;
+        uint8_t num_bits;
+        EventEmitter emitter;
+    };
+
+    struct RangeCheckEntry {
+        uint32_t clk;
+        FF value;
+        uint8_t num_bits;
+        // 8 total 16-bit registers, the last one is dynamic
+        std::array<uint16_t, 7> fixed_slice_registers;
+        uint16_t dynamic_slice_register;
+        // The number of bits that need to be dynamically checked
+        uint16_t dyn_bits;
+        // The difference between
+        uint16_t dyn_diff;
+        // Bit string representing which of the is_lte_x flags are on
+        // From LSB to MSB:
+        // [is_lte_u16, is_lte_u32, is_lte_u48, is_lte_u64, is_lte_u80, is_lte_u96, is_lte_u112, is_lte_u128]
+        uint8_t bit_range_flag;
```

Comment:
Do you need all this or can you have less and syntehsize the rest in finalize()? (see my binary/etc prs)